opal/mca/threads/qthreads/threads_qthreads_mutex.c

```diff
@@ -56,36 +58,81 @@ OBJ_CLASS_INSTANCE(opal_recursive_mutex_t,
                    opal_recursive_mutex_construct,
                    opal_mutex_destruct);
 
-static void opal_cond_create(opal_cond_t *cond)
-{
-    ensure_init_qthreads();
-}
-
 int opal_cond_init(opal_cond_t *cond)
 {
-    return OPAL_ERR_NOT_IMPLEMENTED;
+    opal_atomic_lock_init(&cond->m_lock, 0);
+    cond->m_waiter_head = NULL;
+    cond->m_waiter_tail = NULL;
+    return OPAL_SUCCESS;
 }
 
+typedef struct {
+    int m_signaled;
+    void *m_prev;
+} cond_waiter_t;
+
 int opal_cond_wait(opal_cond_t *cond, opal_mutex_t *lock)
 {
-    ensure_init_qthreads();
-    return 0 == qthread_readFE(*cond, &lock->m_lock_qthreads) ? OPAL_SUCCESS
-                                                              : OPAL_ERROR;
+    opal_threads_ensure_init_qthreads();
+    /* This thread is taking "lock", so only this thread can access this
+     * condition variable.  */
+    opal_atomic_lock(&cond->m_lock);
+    cond_waiter_t waiter = { 0, NULL };
+    if (NULL == cond->m_waiter_head) {
+        cond->m_waiter_tail = (void *)&waiter;
+    } else {
+        ((cond_waiter_t *)cond->m_waiter_head)->m_prev = (void *)&waiter;
+    }
+    cond->m_waiter_head = (void *)&waiter;
+    opal_atomic_unlock(&cond->m_lock);
+
+    while (1) {
+        opal_mutex_unlock(lock);
+        qthread_yield();
+        opal_mutex_lock(lock);
+        /* Check if someone woke me up. */
+        opal_atomic_lock(&cond->m_lock);
+        int signaled = waiter.m_signaled;
+        opal_atomic_unlock(&cond->m_lock);
+        if (1 == signaled)
+            break;
```

Comment:
Nit: please use `{}` around all blocks, even 1-line blocks.

Reply:
I fixed it.  I checked all of my changes and I believe I fixed all of one-liner `if`/`for`/`while`/....